nobody_home.groovy

```diff
@@ -1,164 +1,424 @@
+        state.pendingOp = "init"
+
+        // now set the correct mode for the location. This way, we
+        // don't need to wait for the next sun/presence event.
+
+        // we schedule this action to run after the app has fully
+        // initialized. This way, the app install is faster and the
+        // user customized app name is used in the notification.
+        runIn(7, "setInitialMode")
+    }
+    // On update, we don't change state.modeIfHome. This is so that we
+    // preserve the current sun rise/set state we obtained in earlier
+    // sunset/sunrise handler. This way the app remains in the correct
+    // sun state when the user reconfigures it.
 }
 
-def setSunrise(evt) {
-  changeSunMode(newSunriseMode)
+def setInitialMode()
+{
+    changeSunMode(state.modeIfHome)
+    state.pendingOp = null
 }
 
-def setSunset(evt) {
-  changeSunMode(newSunsetMode)
-}
+// ********** sunrise/sunset handling **********
 
-def changeSunMode(newMode) {
-  state.sunMode = newMode
+// event handler when the sunrise time is reached
+def sunriseHandler(evt)
+{
+    // we store the mode we should be in, IF someone's home
+    state.modeIfHome = newSunriseMode
+    state.currentSunMode = "sunRise"
 
-  if(everyoneIsAway() && (location.mode == newAwayMode)) {
-    log.debug("Mode is away, not evaluating")
-  }
+    // change mode if someone's home, otherwise set to away
+    changeSunMode(newSunriseMode)
+}
 
-  else if(location.mode != newMode) {
-    def message = "${app.label} changed your mode to '${newMode}'"
-    send(message)
-    setLocationMode(newMode)
-  }
+// event handler when the sunset time is reached
+def sunsetHandler(evt)
+{
+    // we store the mode we should be in, IF someone's home
+    state.modeIfHome = newSunsetMode
+    state.currentSunMode = "sunSet"
 
-  else {
-    log.debug("Mode is the same, not evaluating")
-  }
+    // change mode if someone's home, otherwise set to away
+    changeSunMode(newSunsetMode)
 }
 
-def presence(evt) {
-  if(evt.value == "not present") {
-    log.debug("Checking if everyone is away")
-
-    if(everyoneIsAway()) {
-      log.info("Starting ${newAwayMode} sequence")
-      def delay = (awayThreshold != null && awayThreshold != "") ? awayThreshold * 60 : 10 * 60
-      runIn(delay, "setAway")
+def changeSunMode(newMode)
+{
+    // if everyone is away, we need to check and ensure the system is
+    // in away mode.
+    if (isEveryoneAway()) {
+        // this shouldn not happen normally as the mode should already
+        // be changed during presenceHandler, but in case this is not
+        // done, such as when app is initially installed while away,
+        // and system is not in away mode, then we toggle it to away
+        // at the sun rise/set event.
+        changeMode(newAwayMode, " because no one is present")
+    } else {
+        // someone is home, we update the mode depending on
+        // sunrise/sunset.
+        if (state.currentSunMode == "sunRise") {
```

Comment:
Since it's comparison of the same var + a default, a switch statement might be cleaner...but there's only two conditions, so not a concern.